Cryptocell 310 support

RonEld opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->
Add support for Cryptocell  310.
Ported on Nrf52840_DK target.
Adds HW acceleration for the following cryptographic modules:
SHA1, SHA256, AES128_CCM,ECDH, ECDSA

Performance before modification(with ARMCC toolchain):
~~~
  SHA-1                    :       1330 KB/s,         46 cycles/byte
  SHA-256                  :       1158 KB/s,         53 cycles/byte
  AES-CCM-128              :        274 KB/s,        228 cycles/byte
  HMAC_DRBG SHA-1 (NOPR)   :         95 KB/s,        667 cycles/byte
  HMAC_DRBG SHA-1 (PR)     :         88 KB/s,        719 cycles/byte
  HMAC_DRBG SHA-256 (NOPR) :        128 KB/s,        491 cycles/byte
  HMAC_DRBG SHA-256 (PR)   :        113 KB/s,        559 cycles/byte
  ECDSA-secp384r1          :     965 ms/sign
  ECDSA-secp256r1          :     636 ms/sign
  ECDSA-secp384r1          :    1855 ms/verify
  ECDSA-secp256r1          :    1246 ms/verify
  ECDHE-secp384r1          :    1770 ms/handshake
  ECDHE-secp256r1          :    1193 ms/handshake
  ECDHE-Curve25519         :    1055 ms/handshake
  ECDH-secp384r1           :     870 ms/handshake
  ECDH-secp256r1           :     582 ms/handshake
  ECDH-Curve25519          :     542 ms/handshake
~~~
after acceleration:
~~~
  SHA-1                    :      16864 KB/s,          3 cycles/byte
  SHA-256                  :      17385 KB/s,          3 cycles/byte
  AES-CCM-128              :       6886 KB/s,          9 cycles/byte
  HMAC_DRBG SHA-1 (NOPR)   :        286 KB/s,        218 cycles/byte
  HMAC_DRBG SHA-1 (PR)     :        263 KB/s,        238 cycles/byte
  HMAC_DRBG SHA-256 (NOPR) :        420 KB/s,        148 cycles/byte
  HMAC_DRBG SHA-256 (PR)   :        366 KB/s,        170 cycles/byte
  ECDSA-secp384r1          :      46 ms/sign
  ECDSA-secp256r1          :      17 ms/sign
  ECDSA-secp384r1          :      63 ms/verify
  ECDSA-secp256r1          :      22 ms/verify
  ECDHE-secp384r1          :      86 ms/handshake
  ECDHE-secp256r1          :      31 ms/handshake
  ECDHE-Curve25519         :      24 ms/handshake
  ECDH-secp384r1           :      43 ms/handshake
  ECDH-secp256r1           :      16 ms/handshake
  ECDH-Curve25519          :      12 ms/handshake
~~~
Tested also with Mbed TLS self tests
Note: CCM self test passes only with https://github.com/ARMmbed/mbedtls/pull/1598

This is a PR from the private repository - https://github.com/ARMmbed/mbed-os-cc-porting

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [] New target
    [x] Feature
    [ ] Breaking change

Commit:
HW Accelerated SHA1 and SHA256

Add Poritng for Sha1 and SHA256 over Cryptocell

Commit:
Add CCM alternative implementation

Add CCM HW accelerated alternative implementation

Commit:
Alternative ECDSA and ECDH support

Add support for Alternative ECDSA and ECDH, on the higher level,
over CC310. Note that CC generates ECC keys according to FIPS 186,
while mbed TLS generates according to RFC 6979 and RFC 4754,
which causes test vectors for curve p521 to fail

Commit:
Add the Cryptocell library

Add the latest cryptocell library, and add support for CC310 lirary
on latest Mbed OS code

Commit:
Add support for CC trng

add support for HW entropy source

Commit:
Add the platfrom setup \ terminate support

Add support for Platfrom setup and termination for Cryptocell on Nrf52840

Commit:
port CC to IAR

1. Add the compiled CC libraries for IAR toolchain
2. modify `ssi_pal_compiler.h` to support IAR

Commit:
Move trng code from target_nrf5 to target_nrf5x

After rebase, the build target NRF52840_DK is using TARGET_NRF5x,
instead of TARGET_NRF5. Moved Cryptocell TRNG related code from
`targets/TARGET_NORDIC/TARGET_NRF5/` to `targets/TARGET_NORDIC/TARGET_NRF5x/`

Commit:
Rename libraries to pass CI library name check

Rename the CC libraries, with additional prefix "lib*".
Fixes CI failure

RonEld commented:
@0xc0170 How do you suggest I do that? Should I squash all commits into a single one?

RonEld commented:
@0xc0170 I have re-written history, and combined the commits according to topics.

RonEld commented:
Unfortunately, during this rewrite, @k-stachowiak name was removed on some of the contributions

0xc0170 commented on platform/mbed_sdk_boot.c:
```diff
@@ -18,7 +18,9 @@
 #include <stdint.h>
 #include "cmsis.h"
-
+#if DEVICE_CRYPTOCELL
```
This is SDK boot . The mbed OS 5 boot is in rtos folder (unified), see `mbed_boot.c` file

When and how this library should be initialized (why it was added to this boot sequence) ? Why trng_init does not initalizes this (is there better place to do init than adding this here) ?

RonEld replied:
you are correct. This part was already removed as part of 98a9e91db34c8ebad5ad3276097c56ea7d416017

I probably returned it with my history changes. I'll remove it again

0xc0170 reviewed (changes requested):
The git history looks 👍

Commit:
Remove driver intialization from boot code

Remove Cryptocell driver initialization, which was accidently returned
in rebase.

0xc0170 commented on targets/targets.json:
```diff
@@ -3658,6 +3658,9 @@
         "supported_form_factors": ["ARDUINO"],
         "inherits": ["MCU_NRF52840"],
+        "macros_add": ["MBEDTLS_CONFIG_HW_SUPPORT"],
+        "device_has_add": ["CRYPTOCELL"],
```
This is HAL implementation for cryptocell, that we provide. And then what is the extra label for ? To specify the version?

RonEld replied:
I added the `DEVICE_CRYPTOCELL` label for the sake of using the cryptocell trng.
For example, in the platform's [`trng_api.c`](https://github.com/ARMmbed/mbed-os/pull/6794/files#diff-e9952cc585c4b9194267c89d0b9e7390R40), it is mutual to all of NRF5x boards, but on the NRF52840_DK, there is also cryptocell TRNG.
I am thinking that perhaps in the future, to use this label also for the MPS2 porting, done in #6169 but this is only if the API will be common for both targets, which is currently not.

The label for `DEVICE_CRYPTOCELL` is more intended to the partner's code, for them to specify Cryptocell specific function( in case it is common for several boards)

RonEld replied:
If your question is about the `TARGET_CRYPTOCELL310`, this is specifically the CC310 product. Future Cryptocell products will have a different target

0xc0170 replied:
`DEVICE_` is used for HAL, device capabilities, like `trng`. Shouldn't this just be part of it? If cryptocell is enabled, it is used instead ? Does this new cryptocell provide own API or it's just what trng should use? Or does it need better name `CRYPTOCELL_TRNG` ?

I am confused what cryptocell is , why is it being defined as `DEVICE_CRYPTOCELL` and needs extra label of its own to specify a version

RonEld replied:
Cryptocell is Arm's HW accelerator.
perhaps `DEVICE_CRYPTOCELL` is not the right MACRO, but I used `device_has`, same way the MICROVISOR was used.

The extra label, is the actual target version, that partner should add\enable.
The `DEVICE_CRYPTOCELL` is in compilation time, to determine in the partner's common code, for some functionality that should be compiled in \ out of the partner's specific code contains Cryptocell HW. I wouldn't limit it only for TRNG, but this is probably the most common case.
The Cryptocell trng implements same API for trng as any other trng driver, so I needed some special compilation flag to determine the specific board has cryptocell, to avoid multiple definition of same function.

Perhaps the Readme I just added could explain better.

0xc0170 replied:
@sg- Please review (device_has and label addition)

bulislaw replied:
Is cryptocell part of the SoC or is it something separate? That you connect to the board/design?

cmonr replied:
Cryptocell is an ARM security module that is part of the SoC.
The device brief mentions it several times: https://infocenter.nordicsemi.com/pdf/nRF52840_PB_v2.0.pdf

@0xc0170 I think the problem with `CRYPTOCELL_TRNG` is that CRYPTOCELL appears to be more generic than simply applying to the TRNG.

RonEld replied:
Cryptocell is Arm's HW crypto accelerator. It is more than just TRNG, this is why I defined it as `DEVICE_CRYPTOCELL`, to add flexibility for the partner to decide whether to use Cryptocell or their code, if there are duplicates

0xc0170 reviewed (dismissed):
LGTM, just one question about cryptocell versioning via labels.

Commit:
Add guidelines for porting CC310 on Mbed OS

Add guidelines for porting CC310 on a new platform

RonEld commented:
@0xc0170 I eventually uploaded the porting guidelines to this PR. I am not sure if the location is OK.

@AnotherButler Please review the Readme file

Commit:
Copy edit Readme.md

Copy edit file, mostly for minor grammar nits.

AnotherButler commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/Readme.md:
```diff
@@ -0,0 +1,22 @@
+ 	* `MBEDTLS_CONFIG_HW_SUPPORT` to `macros_add` key.
+ 	* `CRYPTOCELL` to `device_has_add` key.
+ 	* `CRYPTOCELL310` to `extra_labels_add` key.
+1. In `objects.h`, include `objects_cryptocell.h`. (You can condition it with `#if DEVICE_CRYPTOCELL` in case you have another `trng` engine for a differnt board, and `objects.h` is common file for your boards, in this case your common `trng_api.c` file should be compiled only if `#if !defined(DEVICE_CRYPTOCELL)`).
```
This line confuses me. Could you please rewrite the second and third sentences for clarity?

AnotherButler reviewed (changes requested):
Please address the comment on line 17.

Commit:
Refactor Readme after comment from Tech Writer

Rephrase usage guideline on `objects.h` and explain usage of every label
in `targets.json`

RonEld commented:
@AnotherButler I have added usage of the labels in `targets.json` and rephrased the sentence accordingly, as you requested

mazimkhan commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/TARGET_NRF52840_DK/cc_platform.h:
```diff
@@ -0,0 +1,33 @@
+typedef struct {
+    char dummy; /**< Placeholder member, as empty structs are not portable. */
+}
+cc_platform_ctx;
```
Can the ```struct``` name be on the same line as closing ```}```.

RonEld replied:
@mazimkhan This is how the Mbed TLS script mandates struct name to be ( I think `check-names.sh`). Although this is MbedOS HAL, and not Mbed TLS, I followed Mbed TLS standards in this part

Patater commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/Readme.md:
```diff
@@ -6,15 +6,15 @@ The CC 310 driver consists of three libraries:
+1. In `objects.h`, include `objects_cryptocell.h`. You can use the `DEBICE_CRYPTOCELL` pre-compilation check as defined above.
 1. In `features/mbedtls/targets/TARGET_CRYPTOCELL310/TARGET_<target name>`, add your platform-specific libraries for all toolchains in `TOOLCHAIN_ARM`, `TOOLCHAIN_GCC_ARM` and `TOOLCHAIN_IAR` respectively.
 1. Add your CC setup code:
 	* Implement `cc_platform_setup()` and `cc_platform_terminate()` to enable CC on your platform, in case you have such limitations. You can implement these functions as empty functions.
```
"in case you have such limitations"

Could you be more explicit in this document about which limitations you have in mind? It's not obvious which limitations are being referred to.

RonEld replied:
perhaps limitations is not the proper wording.  It was meant for board specific setup, prior to CC setup. I'll rephrase

Patater replied:
OK, this is explained well enough now. Thanks.

Commit:
Rephrase explanation on `cc_platform_setup()`

Rephrase the explanation on `cc_platform_setup()`, since previous
explanation wasn't coherent.

mazimkhan commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/Readme.md:
```diff
@@ -0,0 +1,22 @@
+ 	* `MBEDTLS_CONFIG_HW_SUPPORT` to `macros_add` key. This is used to suggest there is HW accelerated cryptography engine that will replace the default Mbed TLS implementation.
+ 	* `CRYPTOCELL` to `device_has_add` key. This should be used in your common code, that you need to remove from compilation in case CC exists in your board. Use `#if !defined(DEVICE_CRYPTOCELL)` or  `#if DEVICE_CRYPTOCELL`.
+ 	* `CRYPTOCELL310` to `extra_labels_add` key. This is used for the build system to look for the CC 310 code and binaries.
+1. In `objects.h`, include `objects_cryptocell.h`. You can use the `DEBICE_CRYPTOCELL` pre-compilation check as defined above.
```
typo DEBICE_CRYPTOCELL -> DEVICE_CRYPTOCELL

mazimkhan commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/aes_alt.c:
```diff
@@ -0,0 +1,1153 @@
+                RK[15] = RK[7] ^ RK[14];
+            }
+            break;
+    }
```
No ```default``` case to handle unexpected values.

RonEld replied:
@mazimkhan This is because in line 722, before this current switch case, we set the value for `ctx->nr`, and it could be only 12 or 14. Otherwise, it owuld return an error - `MBEDTLS_ERR_AES_INVALID_KEY_LENGTH`

Commit:
Copy edit changes to Readme.md

Copy edit changes, mostly for active voice and consistent tense.

AnotherButler reviewed (dismissed):
@RonEld I've completed a final copy edit of this. Please review my changes to make sure I didn't accidentally change the meaning of anything.

hanno-becker commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/cc_rand.h:
```diff
@@ -0,0 +1,73 @@
+}mbedtls_rand_func_container;
+
+
+uint32_t mbedtls_to_cc_rand_func( void *mbedtls_rnd_ctx, uint16_t outSizeBytes, uint8_t *out_ptr )
```
This is defined in a header, yet not declared `static`. Is that deliberate?

hanno-becker replied:
As far as I see, this file is nowhere used. It seems to be a duplicate of code from `cc_internal.h` and `cc_internal.c`.

RonEld replied:
@hanno-arm yes, it should have been removed, and probably returned on my rebase. I'll remove it

RonEld commented:
@AnotherButler Thank you for editing the Readme! I have one question: is "suggests" the correct wording(line 14)? It's not a suggestion, but actually it means that there is HW acceleration code. Perhaps  better wording is:
"This instructs Mbed TLS to look for an alternative cryptographic implementation"?

AnotherButler commented:
@RonEld Ooh, I like that better. 👍

Commit:
Rephrase meaning of `MBEDTLS_CONFIG_HW_SUPPORT`

Rephrase the label, to make it more clearer.

RonEld commented:
@AnotherButler done. Thanks!

Commit:
Resolve issues raised after rebasing

Remove `cc_rand.h` and duplicate check in `mbedtls_ecdsa_sign()`,
that were inserted after rebasing.

Commit:
Fix typo in Readme

Fix typo `DEBICE_CRYPTOCELL ` - > `DEVICE_CRYPTOCELL`

RonEld commented:
@AnotherButler I fixed a typo in the REadme file, found by @mazimkhan 
Please look at last commit

RonEld commented:
@hanno-arm I removed the code accidentally added in rebase. Please look

Commit:
Remove extra `mbedtls_zeroize` call

REmove additional unneded call to `mbedtls_zeroize()` which was added after rebase

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/Readme.md:
```diff
@@ -0,0 +1,22 @@
+
+1. In `targets.json` add the following  to your target:
+ 	* `MBEDTLS_CONFIG_HW_SUPPORT` to `macros_add` key. This instructs Mbed TLS to look for an alternative cryptographic implementation.
+ 	* `CRYPTOCELL` to `device_has_add` key. Use this in your common code that you need to remove from compilation in case CC exists in your board. Use `#if !defined(DEVICE_CRYPTOCELL)` or  `#if DEVICE_CRYPTOCELL`.
```
I dont quite follow this sentence. Do you mean that the port must enclose with those preprocessor directives the code that is disable when CC is in use?

andresag01 replied:
I dont agree with the following sentence:
> use `#if !defined(DEVICE_CRYPTOCELL)` or  `#if DEVICE_CRYPTOCELL`

Note that these two preprocessor directives are NOT equivalent. For example, `#if !defined(DEVICE_CRYPTOCELL)` will succeed if it appears after the line `#define DEVICE_CRYPTOCELL 0` but the check  `#if DEVICE_CRYPTOCELL` will not succeed. I think the usage of these macros should be consistent and that usage should match whatever mbed-cli does to the macros in `device_has_add`.

RonEld replied:
That was my thought, but I tested it, and it works. Note that `DEVICE_CRYPTOCELL` is defined when we have CRYPTOCELL, so I don't see any place where there will be a `#define DEVICE_CRYPTOCELL 0`
So in practice, it's the same. See comment https://github.com/ARMmbed/mbed-os/pull/6794#discussion_r188020917 for more information

andresag01 replied:
All I am suggesting here is to be consistent, so either we should suggest one or the other if it does not make a difference...

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/TARGET_NRF52840_DK/cc_platform.h:
```diff
@@ -0,0 +1,33 @@
+/*
+  *  cc_platform.h
+  *
+  *  Copyright (C) 2018, ARM Limited, All Rights Reserved
```
Minor question: What is the correct format to write the word `Arm`? I thought that its `Arm` (uppercase A, lowercase r and m)

RonEld replied:
You are correct, I'll change it

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/Readme.md:
```diff
@@ -0,0 +1,22 @@
+1. In `objects.h`, include `objects_cryptocell.h`. You can use the `DEVICE_CRYPTOCELL` precompilation check as defined above.
+1. In `features/mbedtls/targets/TARGET_CRYPTOCELL310/TARGET_<target name>`, add your platform-specific libraries for all toolchains in `TOOLCHAIN_ARM`, `TOOLCHAIN_GCC_ARM` and `TOOLCHAIN_IAR` respectively.
+1. Add your CC setup code:
+	* Implement `cc_platform_setup()` and `cc_platform_terminate()` to enable CC on your platform, in case you have board-specific setup functionality, required for CC setup. You can implement these functions as empty functions.
```
I suggest: `You can implement these functions as empty functions.` -> `This functions can be empty`

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/aes_alt.c:
```diff
@@ -0,0 +1,1153 @@
+    *RK++ = *SK++;
+
+exit:
+    memset( &cty, 0, sizeof( struct sw_aes_context ) );
```
I would suggest using a secure memset here.

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/aes_alt.c:
```diff
@@ -0,0 +1,1153 @@
+    if( ctx == NULL )
+        return;
+
+    memset ( ctx, 0, sizeof( mbedtls_aes_context ) );
```
Shouldnt this be a secure memset?

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/aes_alt.h:
```diff
@@ -0,0 +1,120 @@
+    int nr;                     /*!<  number of rounds  */
+    uint32_t *rk;               /*!<  AES round keys    */
+    uint32_t buf[68];           /*!<  unaligned data    */
+};
```
I would suggest using a typedef here and in the next struct.

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/aes_alt.c:
```diff
@@ -0,0 +1,1153 @@
+                    unsigned int keybits )
+{
+    int i, j, ret;
+    struct sw_aes_context cty;
```
I would suggest improving the naming here. For example sw_ctx is more descriptive.

Commit:
Remove AES HW acceleration

Remove AES HW acceleration, as it is targeted future version

Commit:
Remove SW fallback for CCM

Remove SW fallback for CCM mode, and return
`MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE` when trying to set a key
which is not 128 bit size.

RonEld commented:
I have just removed the AES HW acceleration, as it was decided to target it to Mbed OS 5.10.
I removed SW fallback for CCM
Resolved conflicts

Patater commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/objects.h:
```diff
@@ -54,6 +54,14 @@ extern "C" {
 
 #include "nrf_uart.h"
 
+#if DEVICE_CRYPTOCELL
```
Should this be `#if defined(DEVICE_CRYPTOCELL)`?

mazimkhan replied:
mbed-os build system exports some macros with value like ```ABC=1/0```. In that scenario it has to be ```#if ABC``` to work. Though, I am not sure if that is the case here as well.

RonEld replied:
from the compilation, the definition is `-DDEVICE_CRYPTOCELL=1` , so both are valid, because we define `DEVICE_CRYPTOCELL` as 1. I took as example other devices, that are defined this way, such as `#if DEVICE_SERIAL`

hanno-becker replied:
@RonEld If `#if DEVICE_XXX` is what's commonly used, I'd prefer to use that here. As long as both is valid, that's not a blocker, though.

RonEld replied:
@hanno-arm both are used, and after thinking some more, since I haven't tested on NRF5 boards without CRYPTOCELL, I am not sure what the behavior will be if `DEVICE_CRYPTOCELL` is not defined, and we check `#if DEVICE_CRYPTOCELL` .
Also, we use `#if defined(DEVICE_CRYPTOCELL)` in other places in the code, so it's better to be consistent

mazimkhan replied:
Please be aware that it is possible to create a ```mbed_app.json``` in an Mbed OS app where for documentation purpose a user can choose to create a resident macro and only toggle the value ```0/1```. In such scenarios ```#if DEVICE_CRYPTOCELL``` way is better. But if ```DEVICE_CRYPTOCELL``` is defined differently then this does not apply.

>  I am not sure what the behavior will be if ```DEVICE_CRYPTOCELL``` is not defined, and we check ```#if DEVICE_CRYPTOCELL``` .

It would be same as ```#if defined(DEVICE_CRYPTOCELL)```.

mazimkhan replied:
@RonEld my comment is non blocker though.

Patater reviewed (dismissed):
I'm happy with the AES and CCM changes.

I have a question about the CryptoCell 310 Mbed OS defines, specifically about the TRNG.

Otherwise, LGTM.

hanno-becker reviewed (approved):
I approve the CC310 related differences between this PR and the branch https://github.com/ARMmbed/mbed-os-cc-porting/tree/cc310_porting which has been reviewed previously. As @RonEld said, AES H/W acceleration is no longer present, and CCM acceleration is restricted to AES-128 and fails with `FEATURE_UNAVAILABLE` otherwise.

Commit:
Change to DEVICE_CRYPTOCELL definition

Change `#if DEVICE_CRYPTOCELL` to `#if defined(DEVICE_CRYPTOPCELL)`

RonEld commented:
@Patater @hanno-arm I changed `#if DEVICE_CRYPTOCELL` to `#if defined(DEVICE_CRYPTOCELL` per your comments.
We are still waiting for @sg- comment on the ~DEVICE` part, as @0xc0170 mentioned

cmonr commented:
@RonEld Making sure, this is targeting 5.10, correct?

Patater commented:
Hi @cmonr

We are targeting 5.9 with this PR.

RonEld commented:
I would like to summarize current status:
1. The PR has been approved by the Mbed TLS team
2. The Readme has been approved by Tech Writer ( except a small typo fix discovered after)
3. We are waiting @sg- to approve the usage of the `DEVICE_CRYTOCELL`

0xc0170 commented:
Thanks @RonEld 

@bulislaw Can you also review the usage of Cryptocell (how it is defined in hal/targets). here are some details https://github.com/ARMmbed/mbed-os/pull/6794#discussion_r186400242

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/cc_internal.c:
```diff
@@ -0,0 +1,153 @@
+
+int convert_CrysError_to_mbedtls_err( CRYSError_t Crys_err )
+{
+    switch(Crys_err)
```
Minor style: Please put space after ( and before )

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ccm_alt.c:
```diff
@@ -0,0 +1,135 @@
+        return ( MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE );
+    }
+
+    memcpy( ctx->cipher_key , key, keybits/8 );
```
Minor Style: space around arithmetic operators

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ccm_alt.c:
```diff
@@ -0,0 +1,135 @@
+    if( iv_len < 7 || iv_len > 13 )
+        return ( MBEDTLS_ERR_CCM_BAD_INPUT );
+
+    if( length > 0xFFFFFFFF || add_len > 0xFFFFFFFF )
```
I think this is checking if either length or add_lenght are grater than a number that can be encoded in 32 bits. I think for the embedded platforms this check is mostly redundant. While this is harmless, I think that compilers and analysis tools like coverity will raise warnings, so I suggest to at least guard it. See https://github.com/ARMmbed/mbedtls/issues/683

RonEld replied:
I agree, but this was a review comment to verify that length are not bigger than 32 bits, on some platforms. I'll add the check

andresag01 replied:
If you are also targetting platform with more than 32 bits then that makes sense, but as I meant, this could raise warnings in 32 bit platforms because the comparison will never succeed

andresag01 replied:
Also, note that there are several of these comparisons in the PR. I pointed some of them out...

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c:
```diff
@@ -0,0 +1,256 @@
+      if ( pDomain )
+      {
+          uint8_t temp_buf[ 2 * MAX_KEY_SIZE_IN_BYTES + 1 ] = {0};
+          cc_ecc_ws_keygen_params_t* kgParams =  mbedtls_calloc( 1, sizeof(cc_ecc_ws_keygen_params_t) );
```
Minor style: Please put spaces after ( and before )

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c:
```diff
@@ -0,0 +1,256 @@
+          if ( kgParams == NULL )
+              return ( MBEDTLS_ERR_ECP_ALLOC_FAILED );
+          pHeap = kgParams;
+          heapSize = sizeof(cc_ecc_ws_keygen_params_t);
```
Minor style

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c:
```diff
@@ -0,0 +1,256 @@
+          mbedtls_zeroize( temp_buf, sizeof( temp_buf ) );
+      }
+
+      /* if CRYS_ECPKI_GetEcDomain returns NULL, than the given curve is wither Montgomery 25519
```
than -> then

andresag01 replied:
wither -> either

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c:
```diff
@@ -0,0 +1,256 @@
+            goto cleanup;
+        }
+
+        memset ( temp_buf, 0, sizeof(temp_buf)  );
```
Is it necessary to clear the public key information?

RonEld replied:
yes, it is meant to avoid any unneeded information in the buffer, before writing private key information to it. See my answer to you in https://github.com/ARMmbed/mbed-os-cc-porting/pull/7#discussion_r139946103

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c:
```diff
@@ -0,0 +1,256 @@
+        mbedtls_free ( pHeap );
+    }
+
+    if ( secret )
```
I think this check is redundant because it would have returned [at the top](https://github.com/ARMmbed/mbed-os/pull/6794/files#diff-a4b7271ab3959058bdee1757841ea9afR154) if the allocation was not successful.

RonEld replied:
right, but for readability I think it's better this way

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c:
```diff
@@ -0,0 +1,256 @@
+    const CRYS_ECPKI_Domain_t*  pDomain =  CRYS_ECPKI_GetEcDomain ( convert_mbedtls_grp_id_to_crys_domain_id( grp->id ) );
+    uint32_t secret_size =  ( ( grp->nbits + 7 ) / 8 ) ;
+    const uint32_t secret_size_in_heap = secret_size;
+    uint8_t* secret = mbedtls_calloc( 1, secret_size_in_heap );
```
I dont think this will be a problem, but I would like to point out that this is implicitly casting from uint32_t to size_t.

RonEld replied:
I don't follow you. Do you mean `secret_size_in_heap` ? This isn't a problem. `uint32_t` is either same size or smaller than `size_t`

andresag01 replied:
Yes. I dont think its a problem either, but note that `size_t` is not necessarily larger than `uint32_t`. This depends on the system and the C standard says what is the minimum max for size_t (which is smaller than 32 bits), but does not specify a max

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c:
```diff
@@ -0,0 +1,325 @@
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = (unsigned char*)v;
+    while( n-- ) *p++ = 0;
+}
```
This function is repeated across several files. I do not think there will be problems when the new version of Mbed TLS that include mbedtls_platform_zeroize() is used, but mbedtls_zeroize should still be removed.

RonEld replied:
I know. It will be removed when `mbedtls_platform_zeroize()` is introduced.
See https://github.com/ARMmbed/mbed-os-cc-porting/pull/7#discussion_r183620110

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c:
```diff
@@ -0,0 +1,325 @@
+    mbedtls_rand_func_container cc_rand = { f_rng, p_rng };
+    const CRYS_ECPKI_Domain_t*  pDomain =  CRYS_ECPKI_GetEcDomain ( convert_mbedtls_grp_id_to_crys_domain_id( grp->id ) );
+
+    if( blen > 0xFFFFFFFF )
```
Same issue with comparison as mentioned above.

RonEld replied:
ok

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c:
```diff
@@ -0,0 +1,325 @@
+    if( blen > 0xFFFFFFFF )
+    {
+        ret = MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+        goto cleanup;
```
You could just return here since we have not allocated or done anything yet

RonEld replied:
ok, probably a copy \ paste error. But this is not blocking, in my opinion

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c:
```diff
@@ -0,0 +1,325 @@
+    uint32_t signature_size =  ( ( grp->nbits + 7 ) / 8 ) *2;
+    const uint32_t signature_size_for_heap = signature_size;
+    mbedtls_rand_func_container cc_rand = { f_rng, p_rng };
+    const CRYS_ECPKI_Domain_t*  pDomain =  CRYS_ECPKI_GetEcDomain ( convert_mbedtls_grp_id_to_crys_domain_id( grp->id ) );
```
I would suggest moving the assignment to until after the parameter checks.

RonEld replied:
It's more efficient to initialize a variable with a  value, rather than assign it later.
Nonetheless, I don't think this is a blocker

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c:
```diff
@@ -0,0 +1,325 @@
+
+    if ( pDomain != NULL )
+    {
+        uint8_t temp_buf[ MAX_KEY_SIZE_IN_BYTES ] = {0};
```
I would suggest just moving the declaration of this buffer to the top of the function and adding a memset there to avoid having scattered calls to mbedtls_zeroize()  through the body of the function. However, having said that, I cannot find any issues with the current implementation either, so this is just a suggestion

RonEld replied:
This `temp_buf` is relevant only for WS curves, which are handled in this scope. I prefer not to have this buffer if not used

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c:
```diff
@@ -0,0 +1,325 @@
+    uint32_t signature_size = ( ( grp->nbits + 7 ) / 8 ) * 2;
+    const CRYS_ECPKI_Domain_t*  pDomain =  CRYS_ECPKI_GetEcDomain ( convert_mbedtls_grp_id_to_crys_domain_id( grp->id ) );
+
+    if( blen > 0xFFFFFFFF )
```
Same as before.

RonEld replied:
ok, probably a copy \ paste error. But this is not blocking, in my opinion

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c:
```diff
@@ -0,0 +1,325 @@
+        if ( ret != 0 )
+            goto cleanup;
+
+        memset ( temp_buf, 0 , sizeof(temp_buf) );
```
Is clearing this buffer necessary? If it is not, I would suggest removing this line, but if it is I suggest using mbedtls_zeroize as memset could be optimised away (though its very unlikely in this case.

RonEld replied:
yes. See my answer in https://github.com/ARMmbed/mbed-os/pull/6794#discussion_r188532708
It is very unlikely in this case that it will be optimized away, if there is actual data in the buffer

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/platform_alt.c:
```diff
@@ -0,0 +1,57 @@
+ * Until then, we should keep it global and extern */
+
+CRYS_RND_State_t     rndState = { { 0 } } ;
+CRYS_RND_WorkBuff_t  rndWorkBuff = { { 0 } } ;
```
Question: I think I understand what the problem is with the setters/getters, but why is it not possible to just put these two into the platform context and keep that global and extern?

RonEld replied:
using the platform context is not according to design. 
`rndWorkBuff` is just a temporary sandbox, and wouldn't be in the platform context anyway
If we put the platform context as global, it is more likely it will be forgotten this way, rather than put `rndState` as global. If we put the platform context as global, having it as parameter is redundant

andresag01 replied:
I dont follow. You quoted an issue that says we do not have a clean way of accessing the platform context from the hw acceleration code, but if we did then you would put rndState there. Do you mean that you do not want to make the platform context global? But if there was some sort of getter like suggested in the issue it would be "global" in a way. I dont really see why this is not in the platform context, but I am probably misunderstanding something

RonEld replied:
Well, we haven't thought about the architecture yet.l Anyway, setter and getter doesn't make the context global, and keeps the abstraction of the context.

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/sha1_alt.c:
```diff
@@ -0,0 +1,86 @@
+void mbedtls_sha1_free( mbedtls_sha1_context *ctx )
+{
+    if( ctx == NULL )
+        return;
```
Why put this check here and not in mbedtls_sha1_init() as well?

RonEld replied:
This is the same behavior as in our SW code

andresag01 replied:
ok

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/sha1_alt.c:
```diff
@@ -0,0 +1,86 @@
+    if( CrysErr == CRYS_OK )
+    {
+        memcpy( output, crys_result, 20 );
+        return ( 0 );
```
Question: Should we zeroize crys_result before returning?

RonEld replied:
I don't think so. `crys_result` is not a secret

andresag01 replied:
I thought that crys_result was the hash of the data that you gave this module, right? If that is the case, then crys_result can be the hash of sensitive data like a password, etc...

RonEld replied:
`crys_result` is the hash. but it is also the output of the function. The hash of sensitive data is not a secret

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/sha1_alt.c:
```diff
@@ -0,0 +1,86 @@
+    return ( 0 );
+}
+
+int mbedtls_sha1_finish_ret( mbedtls_sha1_context *ctx,
```
Question: What happens if someone calls finish right after starts (no update in between)? Does this cause a failure? I remember one of the partners had issues with this when implementing their hardware acceleration as one of the tests did just that.

RonEld replied:
shouldn't cause an issue

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/sha256_alt.c:
```diff
@@ -0,0 +1,86 @@
+/*
+ *  sha256_alt.c
```
Same questions/comments in this module as with SHA1

RonEld replied:
same answers

andresag01 commented on targets/TARGET_CRYPTOCELL310/objects_cryptocell.h:
```diff
@@ -0,0 +1,44 @@
+    uint32_t  SubSamplingRatio3;
+    uint32_t  SubSamplingRatio4;
+
+}CRYS_RND_Params_t;
```
Should this be enclosed in something like `#if defined(DEVICE_TRNG) ... #endif`?

RonEld replied:
no. I took example from other `objects.h` files. This is only a definition of structs, and having this file included implies there is `DEVICE_TRNG`

andresag01 replied:
ok

andresag01 reviewed (commented):
The changes are looking good. I just left minor comments and questions.

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/sha1_alt.c:
```diff
@@ -0,0 +1,86 @@
+
+    CRYS_HASH_Free( &ctx->crys_hash_ctx );
+
+    memset( ctx, 0, sizeof( mbedtls_sha1_context ) );
```
Should this be mbedtls_platform_zeroize()? That is what mbedtls_sha1_free() does: https://github.com/ARMmbed/mbedtls/blob/development/library/sha1.c#L84

Commit:
Minor style fixes

Fix styls issues and rename ARM to Arm.
Address additional minor comments from review.

RonEld commented:
@andresag01 I addressed your issues. Note that the aes HW acceleration was removed from this PR, so I disregarded your comments on it.
@AnotherButler Could you please review the final Readme file, after typo fix and small rephrase?

hanno-becker commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/Readme.md:
```diff
@@ -12,11 +12,11 @@ To port your CC 310 driver to Mbed OS on your specific target, do the following:
 1. In `features/mbedtls/targets/TARGET_CRYPTOCELL310/TARGET_<target name>`, add your platform-specific libraries for all toolchains in `TOOLCHAIN_ARM`, `TOOLCHAIN_GCC_ARM` and `TOOLCHAIN_IAR` respectively.
 1. Add your CC setup code:
-	* Implement `cc_platform_setup()` and `cc_platform_terminate()` to enable CC on your platform, in case you have board-specific setup functionality, required for CC setup. You can implement these functions as empty functions.
+	* Implement `cc_platform_setup()` and `cc_platform_terminate()` to enable CC on your platform, in case you have board-specific setup functionality, required for CC setup.These functions can be empty.
```
A space is missing before `There`.

hanno-becker reviewed (changes requested):
A space went missing in the final commit, otherwise it looks good to me.

andresag01 commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c:
```diff
@@ -175,11 +176,12 @@ int mbedtls_ecdsa_verify( mbedtls_ecp_group *grp,
     uint32_t signature_size = ( ( grp->nbits + 7 ) / 8 ) * 2;
     const CRYS_ECPKI_Domain_t*  pDomain =  CRYS_ECPKI_GetEcDomain ( convert_mbedtls_grp_id_to_crys_domain_id( grp->id ) );
 
+#if defined(MBEDTLS_HAVE_INT64)
```
Please note that `MBEDTLS_HAVE_INT64` does not mean that size_t is 64 bits. It simply checks whether the underlying platform supports uint64_t. We had issues because of this in the past: https://github.com/ARMmbed/mbedtls/pull/1043

Commit:
Add space

Add space which was lost in previous commit

RonEld commented:
Status update:
1. After @andresag01 comments were addressed, This PR has been re-approved by the Mbed TLS Team
2. The Readme has been slightly modified. I think we should have @AnotherButler review it again.
3. Waiting for review from @sg- or @bulislaw about the `DEVICE_CRYPTOCELL` usage

bulislaw commented on features/mbedtls/targets/TARGET_CRYPTOCELL310/Readme.md:
```diff
@@ -0,0 +1,22 @@
+# Guidelines for porting Cryptocell to Mbed OS
```
That should go to the Handbook repo.

RonEld replied:
I don't think so. This is not a generic porting guide. Quoting @AnotherButler about this (through mail):
>. We don’t include board- or module-specific information in the Handbook. However, if there’s porting information that’s applicable to all boards or modules, we’ll happily put that in the Handbook. If it’s board-specific, let’s keep it in the README. 

This is not board specific but it's component specific

bulislaw reviewed (commented):
I've noticed you add crypto cell as a target, I'm don't think that's correct. We are not going to run Mbed OS on cryptocell we will be merely using it from Mbed OS from my understanding.

RonEld commented:
@bulislaw Thank you for your comments.
Cryptocell is Arm's HW accelerator. It is not exactly a target, as it's a component added to the target. I did consider having it a `FEATURE` rather than a `TARGET` but all the other board HW accelerators are put as targets. It is part of the SoC, as @cmonr mentioned. You will also find description for it in  [here](https://developer.arm.com/products/system-ip/trustzone-security-ip)
How do you suggest we integrate it, and keep the current architecture of all HW accelerators drivers are located in `features/mbedtls/targets/` ?

bulislaw commented:
> all the other board HW accelerators are put as targets.

@RonEld It's not really accurate looking at https://github.com/ARMmbed/mbed-os/tree/master/features/mbedtls/targets/TARGET_STM and others we don't add `TARGET_accelerator` but rather just extend existing targets to support the accelerators. @Patater
 
I'm not sure how would that work to have two separate targets in one build. I think it's more of a feature  of a target than the target itself.

RonEld commented:
@bulislaw 
> It's not really accurate looking at https://github.com/ARMmbed/mbed-os/tree/master/features/mbedtls/targets/TARGET_STM and others we don't add TARGET_accelerator but rather just extend existing targets to support the accelerators

Correct, but that's basically because the accelerators are supplied by the target developer, and it's part of the target code. in this case, the accelerator is not entirely part of the target. 

>I'm not sure how would that work to have two separate targets in one build.

Well, in this specific PR, we already have two separate targets, `TARGET_NRF53840_DK` and `TARGET_CRYPTOCELL310`, and it works fine. the build system searches for the code in both folders. (This is why I had to add the `DEVICE_CRYPTOCELL` label for places where there are duplicate functions). I did consider having it as a feature, but I am not sure it worked, and it clashes with the design that the accelerator code should be in the `mbedtls/targets` folder. I don't mind checking again if adding it as a feature will work, but I think it shouldn't block this PR, and we can do it in a future PR.

bulislaw commented:
It doesn't fit in what we define as targets in Mbed OS and it doesn't fit our config system (hence target and device). It does fit exactly what we define as a feature. I don't feel comfortable merging it as it is. Shouldn't be too much work to rename the defines to only have feature and not TARGET and DEVICE.

bulislaw reviewed (changes requested):
Please integrate cryptocell as a feature not a target.

RonEld commented:
@bulislaw It's not a matter of renaming defines, it's a matter of the build system knowing to search in correct location, and having the crypto accelerator code outside of the `mbedtls` folder.
I will work on integrating it as a feature

bulislaw commented:
Sorry for being a pain, but right now it really doesn't fit our definitions of things.

RonEld commented:
No problem. where do you recommend putting the TRNG driver? In the HAL, it should be part of `targets` folder.

Commit:
Change Cryptocell target to a feature

Change the Cryptocell310 target to `FEATURE_CRYPTOCELL310`

RonEld commented:
@bulislaw I change `TARGET_CRYPTOCELL310` to `FEATURE_CRYPTOCELL310` and modified files and location accordingly.
Please review.
I would like to state that I less like this approach, as it will make us need to update `tools/config/__init__.py` whenever we want to add a new CryptoCell 3XX family version as a supported feature

adbridge commented:
@bulislaw are you happy with this update ?

RonEld commented:
Note that the CI fails on doxygen of third party code ( IPG code )

marcuschangarm commented on targets/targets.json:
```diff
@@ -3682,6 +3682,8 @@
         "supported_form_factors": ["ARDUINO"],
         "inherits": ["MCU_NRF52840"],
+        "macros_add": ["MBEDTLS_CONFIG_HW_SUPPORT"],
+        "features": ["CRYPTOCELL310"],
```
This should go in the `MCU_NRF52840` section since it is MCU specific and not target specific.

marcuschangarm commented:
Since this feature is MCU specific, the NRF52840 target should be `TARGET_MCU_NRF52840` and not the development kit target.

The target specific files should then go in the folder: 
`features/cryptocell/FEATURE_CRYPTOCELL310/TARGET_MCU_NRF52840`

Commit:
Change target name

Apply the CC changes to TARGET_MCU_NRF52840 , and not to
the DK target alone

RonEld commented:
@marcuschangarm Thank you for your coment. I have change to the `TARGET_MCU_NRF52840` target, although the only target I found could be built is the DK target.

RonEld commented:
Current status:
1. All changes approved by Mbed TLS team
2. Changes in Readme file *may* need final review from Tech Author. @AnotherButler any thoughts?
3. Changes of CC from DEVICE \ target to feature pending review from @bulislaw 
<del>4. CI fails because of doxygen script, in the header files of third party package. What can be done on this?</del>

Commit:
Avoid doxyfile checks on cryptocell feature

Exclude cryptocell feature in the doxyfile script

0xc0170 reviewed (dismissed):
One question:

files in `/features/cryptocell/` are mbedtls? If yes, astyleignore file should contain this folder to ignore doing style check, otherwise if not, then they should conform to our coding style.

RonEld commented:
@0xc0170 It's not mbedtls, but it is third party.
I'll add it to astyleignore as well

Commit:
ignore cryptocell in astyle check

Add cryptocell to `astyleignore`

Patater commented on .astyleignore:
```diff
@@ -1,5 +1,6 @@
 BUILD
 cmsis
+/features/cryptocell
```
This shouldn't have a `/` on the front, should it?

Patater reviewed (dismissed):
Other than the possible astyle problem, LGTM.

yanesca reviewed (dismissed):
LGTM

RonEld commented:
@Patater yes, it should be without the `/`. I amended the commit.
Unfortunately, your (and @yanesca ) approval was dismissed

0xc0170 commented:
@Patater @yanesca @bulislaw Please approve. This feature looks to me complete

marcuschangarm commented:
> I have change to the TARGET_MCU_NRF52840 target, although the only target I found could be built is the DK target.

Thank you! The DK is the only one at the moment, but any future boards should be inheriting from MCU_NRF52840 so now we are future ready.

Commit:
Check the size_t isn't larger than 32 bit

Assure that `size_t` isn't larger than 32 bit, with preprocessor check.
Using `#if SIZE_MAX > UINT_MAX`.

RonEld commented:
I addressed [comment](https://github.com/ARMmbed/mbed-os/pull/6794#discussion_r188595731) from @andresag01 regarding assuring the `size_t` is not 64 bit. 

Current status:
1. Last minor change awaiting review from @Patater and @yanesca 
2. Readme possible awaiting review from Tech writer for the last modifications
3. waiting review from @bulislaw

0xc0170 commented:
/morph build

cmonr commented:
Shuffling around some jobs.

/morph export-build

cmonr commented:
/morph test

cmonr commented:
Log saved for further debugging. Restarting test since failure seems timing/CI load related.

/morph test
